### PR TITLE
[REVIEW] Add full MLflow + k8s example on GCP

### PR DESCRIPTION
**Pull request by drobison00**

### CODE REVIEW

File: mlflow_project/docker_environment/src/k8s/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
Comment (JohnZed) [inline mlflow_project/docker_environment/src/k8s/entrypoint.sh:1]: A quick header comment in each entrypoint explaining roughly what it does (like a one liner) would be ahndy

File: mlflow_project/docker_environment/src/rf_test/train_simple.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    mlflow.log_metric("accuracy", acc)
+
+    mlflow.sklearn.log_model(mod, "saved_models")
+
+    return mod
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
```
Comment (JohnZed) [inline mlflow_project/docker_environment/src/rf_test/train_simple.py:64]: I'm not quite clear about how train and train_simple relate or differ?

File: mlflow_project/local_environment/envs/conda.yaml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  - heapdict=1.0.1=py_0
+  - icu=64.2=he1b5a44_1
+  - jinja2=2.11.2=pyh9f0ad1d_0
+  - joblib=0.15.1=py_0
+  - jpeg=9d=h516909a_0
+  - ld_impl_linux-64=2.33.1=h53a641e_7
+  - libblas=3.8.0=16_openblas
+  - libcblas=3.8.0=16_openblas
+  - libcudf=0.14.0=cuda10.2_0
+  - libcuml=0.14.0=cuda10.2_0
```
Comment (JohnZed) [inline mlflow_project/local_environment/envs/conda.yaml:48]: can we bump all to 0.15?

File: mlflow_project/local_environment/src/rf_test/train.py [hunk 4]
```diff
@@ -76,47 +78,38 @@ def train(params, fpath, hyperopt=False):
         return _train(params, fpath, hyperopt)
 
 
-if (__name__ == "__main__"):
+if __name__ == "__main__":
```
Comment (JohnZed) [inline mlflow_project/local_environment/src/rf_test/train.py:81]: how do these train files relate to train and train_simple above in docker_environment?

**Review by JohnZed — CHANGES_REQUESTED** [review]
Phew! That's a big one... Looks very cool. My main suggestion would be that it could use a higher-level explanation doc that describes roughly what it's doing and what the steps are at a 30,000 foot view. There are a lot of files and pieces, and the docs of specific steps seem fairly clear, but it would be hard for someone to just pull up a page and know what's going on without more background. Additionally, it would be nice to make it clear how a user can easily drop in and iterate on their own code for mlflow here - I think it's fairly straightforward, but that would be nice to spell out.

File: mlflow_project/docker_environment/Dockerfile.training [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+RUN source activate rapids \
+    && mkdir /opt/mlflow \
+    && pip install \
+        boto3 \
+        hyperopt \
+        mlflow \
+	    psycopg2-binary
+
+COPY airline_small.parquet /rapids/data/airline_small.parquet
```
[the remark this answers is not in the document]
Comment (drobison00) [reply mlflow_project/docker_environment/Dockerfile.training:11]: I agree; however, doing it the correct way, means we now we have to create persistent storage, with the data populated, and attach it to the pod we're deploying. Which adds a number of additional steps to the process that aren't directly related to RAPIDS or MLflow.

File: mlflow_project/docker_environment/src/rf_test/train_simple.py (hunk 2 above)
Comment (drobison00) [reply mlflow_project/docker_environment/src/rf_test/train_simple.py:64]: train.py is MLflow + hyperopt, train_simple.py is just MLflow integration.

File: mlflow_project/local_environment/envs/conda.yaml (hunk 3 above)
Comment (drobison00) [reply mlflow_project/local_environment/envs/conda.yaml:48]: Yep. I'll regenerate these and retest.

File: mlflow_project/docker_environment/README.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    - Training Data
+        - Download:
+            ```shell script
+            wget -N https://rapidsai-cloud-ml-sample-data.s3-us-west-2.amazonaws.com/airline_small.parquet
+            ```
+        - Push to your storage bucket
+            - We'll assume that it lives at GS_DATA_PATH: `gs://${YOUR_BUCKET}/airline_small.parquet`
+    - Environment Setup 
+        - Upload envs/conda.yaml to your GCP cloud storage bucket
+            - We'll assume that it lives at GS_CONDA_PATH: `gs://${YOUR BUCKET}/conda.yaml`
```
Comment (Nanthini10) [inline mlflow_project/docker_environment/README.md:76]: Variable name missing an underscore`YOUR_BUCKET`

File: mlflow_project/docker_environment/README.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        helm install mlf-db bitnami/postgresql --set postgresqlDatabase=mlflow_db --set postgresqlPassword=mlflow \
+          --set service.type=NodePort
+        ```
+- **Tracking Server deployment**
+    - Build a docker container that will host our MLflow [Tracking Service](https://www.mlflow.org/docs/latest/tracking.html#mlflow-tracking-servers)
+    and publish it to our GCR repository. The tracking service will provide a REST interface exposed via kubernetes load
+    balancer, which should be the target of our **MLFLOW_TRACKING_URI** variable.
+        ```shell script
+        docker build --tag ${GCR_REPO}/mlflow-tracking-server:gcp --file Dockerfile.tracking .
+        docker push ${GCR REPO}/mlflow-tracking-server:latest
```
Comment (Nanthini10) [inline mlflow_project/docker_environment/README.md:115]: Missing an underscore `GCR_REPO`

File: mlflow_project/docker_environment/README.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+**Create and publish a training container**
+- **Build a training container**.
+    - This creates the base docker container that MLflow will inject our project into, and deploy into our Kubernetes cluster
+    for training.
+    - `docker build --tag rapids-mlflow-training:gcp --file Dockerfile.training .`
+
+**Configure your MLProject to use Kubernetes**    
+- **Export MLFLOW_TRACKING_URI**
+    - This should be the REST endpoint exposed by our tracking server (see above).
+    - `export MLFLOW_TRACKING_URI=http://${MLFLOW TRACKING_URI}`
```
Comment (Nanthini10) [inline mlflow_project/docker_environment/README.md:158]: Missing an underscore too - `MLFLOW_TRACKING_URI`

File: mlflow_project/docker_environment/README.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    ```
+- **Log in to your tracking server and verify that the experiment was successfully logged, and that you have a registered
+    model**.
+    ![](images/ts_single_run.png)
+    ![](images/ts_registered_model.png)
+    
+- **Serve a trained model, locally**
+    - Set you service account credentials
+        ```shell script
+        export MLFLOW_TRACKING_URI=http://${MLFLOW TRACKING_URI}
```
Comment (Nanthini10) [inline mlflow_project/docker_environment/README.md:182]: Same as before `MLFLOW_TRACKING_URI`

**Review by Nanthini10 — CHANGES_REQUESTED** [review]
Looks good to me. I was able to run this fully. Just a few typos in variable names.

**Review by jacobtomlinson — COMMENTED** [review]
I've had a read through and this all looks great. 

A nice addition might be to use [Frigate](https://frigate.readthedocs.io/en/latest/) to generate a README to put in `
cloud-ml-examples/mlflow_project/docker_environment/helm/mlflow-tracking-server/
`. Frigate takes in the info in `Chart.yaml` and renders a README from it and includes a table of config options generated from `values.yaml`. You can optionally take it a step further by annotating the `values.yaml` file with comments too to describe what each option does.

**Pull request merged**